source/extensions/filters/http/wasm/wasm_filter.cc

```diff
@@ -46,6 +31,16 @@ FilterConfig::FilterConfig(const envoy::extensions::filters::http::wasm::v3::Was
   }
 }
 
+FilterConfig::~FilterConfig() {
+  if (tls_slot_->currentThreadRegistered()) {
```

Comment:
Let's focus on this one:

At startup:
```
main thread     -> [39289][warning][wasm] [source/extensions/filters/http/wasm/wasm_filter.cc:26] XXX before createWasm
main thread     -> [39289][warning][wasm] [source/extensions/filters/http/wasm/wasm_filter.cc:21] XXX before getOrCreateThreadLocalWasm
worker thread 1 -> [39577][warning][wasm] [source/extensions/filters/http/wasm/wasm_filter.cc:21] XXX before getOrCreateThreadLocalWasm
worker thread 2 -> [39578][warning][wasm] [source/extensions/filters/http/wasm/wasm_filter.cc:21] XXX before getOrCreateThreadLocalWasm
```

With config change at runtime:
```
main thread     -> [39289][warning][wasm] [source/extensions/filters/http/wasm/wasm_filter.cc:26] XXX before createWasm
main thread     -> [39289][warning][wasm] [source/extensions/filters/http/wasm/wasm_filter.cc:21] XXX before getOrCreateThreadLocalWasm
worker thread 2 -> [39578][warning][wasm] [source/extensions/filters/http/wasm/wasm_filter.cc:21] XXX before getOrCreateThreadLocalWasm
worker thread 1 -> [39577][warning][wasm] [source/extensions/filters/http/wasm/wasm_filter.cc:21] XXX before getOrCreateThreadLocalWasm
main thread     -> [39289][warning][wasm] [source/extensions/filters/http/wasm/wasm_filter.cc:37] XXX before if currentThreadRegistered
main thread     -> [39289][warning][wasm] [source/extensions/filters/http/wasm/wasm_filter.cc:40] XXX inside the runOnAllThreads callback
worker thread 1 -> [39577][warning][wasm] [source/extensions/filters/http/wasm/wasm_filter.cc:40] XXX inside the runOnAllThreads callback
worker thread 2 -> [39578][warning][wasm] [source/extensions/filters/http/wasm/wasm_filter.cc:40] XXX inside the runOnAllThreads callback
```

After `/quitquitquit`:
```
main thread     -> [39289][warning][wasm] [source/extensions/filters/http/wasm/wasm_filter.cc:37] XXX before if currentThreadRegistered
main thread     -> [39289][critical][assert] [source/common/thread_local/thread_local_impl.cc:150] assert failure: !shutdown_.
```

Reply:
OK, I've looked at all the other instances of `runOnAllThreads()` and it doesn't look that any of the other extensions use it for cleanup during shutdown. What's more, `ThreadLocalStoreImpl` specifically checks for `!shutting_down_` before executing anything via `runOnAllThreads()`, so I don't think it's supposed to work during shutdown.

Should I add a comment to say that cleanup is skipped during shutdown? Or do you want me to make some other changes?

Reply:
There are reports of `assert failure: currentThreadRegisteredWorker(index).` in the upcoming Istio release (based on Envoy 1.16.0), so perhaps there is a recent regression:
```
assert failure: currentThreadRegisteredWorker(index).
Caught Aborted, suspect faulting address 0x53900000012
Backtrace (use tools/stack_decode.py to get line numbers):
Envoy version: 211ca038383b61c9fb600758bf170aea5508800c/1.16.0/Clean/DEBUG/BoringSSL
 #0: Envoy::SignalAction::sigHandler() [0x5602aeb3f66c]
 #1: __restore_rt [0x7fbabc9518a0]
 #2: Envoy::ThreadLocal::InstanceImpl::SlotImpl::runOnAllThreads()::$_2::operator()() [0x5602ada1bffc]
 #3: std::__1::__invoke<>() [0x5602ada1bfad]
 #4: std::__1::__invoke_void_return_wrapper<>::__call<>() [0x5602ada1bf5d]
 #5: std::__1::__function::__alloc_func<>::operator()() [0x5602ada1bf2d]
 #6: std::__1::__function::__func<>::operator()() [0x5602ada1b14e]
 #7: std::__1::__function::__value_func<>::operator()() [0x5602a92053c5]
 #8: std::__1::function<>::operator()() [0x5602a9205385]
 #9: Envoy::ThreadLocal::InstanceImpl::SlotImpl::wrapCallback()::$_0::operator()() [0x5602ada19046]
#10: std::__1::__invoke<>() [0x5602ada18fed]
#11: std::__1::__invoke_void_return_wrapper<>::__call<>() [0x5602ada18f9d]
#12: std::__1::__function::__alloc_func<>::operator()() [0x5602ada18f6d]
#13: std::__1::__function::__func<>::operator()() [0x5602ada1814e]
#14: std::__1::__function::__value_func<>::operator()() [0x5602a92053c5]
#15: std::__1::function<>::operator()() [0x5602a9205385]
#16: Envoy::ThreadLocal::InstanceImpl::runOnAllThreads() [0x5602ada143f1]
#17: Envoy::ThreadLocal::InstanceImpl::SlotImpl::runOnAllThreads() [0x5602ada13a97]
#18: Envoy::Extensions::Clusters::Aggregate::Cluster::refresh() [0x5602a95817de]
...
```
(From: https://github.com/istio/istio/issues/28620)

Reply:
RE cleanup during shutdown: why not just rely on the destructors for the objects held in the TLS slots?

Reply:
So, in the "early times" IIRC there used to be a shutdown() method for TLS objects that we would call to avoid this type of situation. Essentially we would give each TLS object a chance to clean itself up. I don't remember the history, but I removed it at some point because I determined all of the lifecycle stuff could be simplified because we destroy all TLS objects when shutdownThread() is called. So per @jmarantz why can't the TLS object constructor just do what it has to do when it's destroyed on the thread its owned by?

For the aggregate cluster issue, it's possible there is a regression from some of the ref counting changes, but I think this is unrelated. Can you or someone else open a different issue on that with a complete callstack and whether it's happening during shutdown or not?

Reply:
We're doing graceful shutdown of the Proxy-Wasm plugins within the WasmVM, so we cannot destroy them in-place, but I think that if I hold onto thread-local object representing RootContext instead of shared Plugin, then I might be able to dispatch this without `runOnAllThreads`. Let me see what I can do.